[css-animations] resolve `animation-name` as tree-scoped references

graouts opened the pull request:
#### 5e9cf5c17d3448f24c84ba5af5529c69c0fc9935
<pre>
[css-animations] resolve `animation-name` as tree-scoped references
<a href="https://bugs.webkit.org/show_bug.cgi?id=296048">https://bugs.webkit.org/show_bug.cgi?id=296048</a>
<a href="https://rdar.apple.com/156484228">rdar://156484228</a>

Reviewed by Antti Koivisto.

Adopt `Style::Scope::resolveTreeScopedReference()` introduced in 300333@main to resolve `animation-name`
and find the correct scope to find the matching `@keyframes` rule.

Because `resolveTreeScopedReference()` provides a const `Style::Scope` and `Style::Resolver` to work with,
a few methods on `Style::Resolver` needed to be updated to be marked const, which they should have been
all along.

Additionally, because the `resolveTreeScopedReference()` lambda needs to return a truthy value to determine
whether a matching `@keyframes` rule was found for the provided scope, we needed to change
`Style::Resolver::keyframeStylesForAnimation()` to return a `bool` to indicate whether it processed an
`@keyframes` rule.

Tests: imported/w3c/web-platform-tests/css/css-animations/animation-name-in-nested-shadow.html
       imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-inner-match.html
       imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-outer-match.html
       imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part.html

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-nested-shadow-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-nested-shadow-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-nested-shadow.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-inner-match-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-inner-match.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-outer-match-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-outer-match.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/keyframes-002-expected.txt:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::computeCSSAnimationBlendingKeyframes):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::builderContext const):
(WebCore::Style::Resolver::styleForKeyframe const):
(WebCore::Style::Resolver::isAnimationNameValid const):
(WebCore::Style::Resolver::keyframeStylesForAnimation const):
(WebCore::Style::Resolver::builderContext): Deleted.
(WebCore::Style::Resolver::styleForKeyframe): Deleted.
(WebCore::Style::Resolver::isAnimationNameValid): Deleted.
(WebCore::Style::Resolver::keyframeStylesForAnimation): Deleted.
* Source/WebCore/style/StyleResolver.h:
* Source/WebCore/style/Styleable.cpp:
(WebCore::keyframesRuleExistsForAnimation):

Canonical link: <a href="https://commits.webkit.org/300706@main">https://commits.webkit.org/300706@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e11fc927a7f452b5ca8233d5836425a3b9788e6c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43280 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33976 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130295 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75708 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125442 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44003 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51874 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93937 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62361 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126518 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35055 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110545 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74555 "Found 1 new API test failure: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlot (failure)") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34025 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28703 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73808 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104773 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28927 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133019 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50516 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38458 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102431 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50891 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106766 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102273 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47623 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25860 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19452 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50370 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56131 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49844 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53191 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51519 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleScope.cpp:
```diff
@@ -257,6 +258,37 @@ const Scope* Scope::forOrdinal(const Element& element, ScopeOrdinal ordinal)
+    // tree-scoped names "inherit" into descendant shadow trees, so long as they don’t define
+    // the same name themselves.)
+
+    if (element.containingShadowRoot()) {
+        auto& styleScope = forNode(element);
+        if (styleScope.resolver().isAnimationNameValid(scopedName.name))
+            return &styleScope;
+    }
+
+    if (auto* styleScope = forOrdinal(element, scopedName.scopeOrdinal)) {
+        if (styleScope->resolver().isAnimationNameValid(scopedName.name))
+            return styleScope;
```
Shouldn't these be other way round: "first search only the tree-scoped names associated with the same root"?

Also it should probably be searching the entire containingShadowRoot chain in a loop rather than just the closest root.
.

anttijk reviewed (changes requested):
I can't actually map this code to the quoted spec test.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/510288fff8c1b8c604f1c89bed3261c068b7d705. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50302

anttijk commented on LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-inner-match.html:
```diff
@@ -0,0 +1,47 @@
+<!DOCTYPE html>
+<html>
+<head>
+<title>Matching animation-name from within a shadow root with multiple @keyframes rules</title>
```
These tests are not as comprehensive as css/css-anchor-position/position-try-tree-scoped.html which also tests nested shadow trees.

graouts replied:
I added a fourth test which show nesting similar to `css/css-anchor-position/position-try-tree-scoped.html`.

graouts committed:
[css-animations] resolve `animation-name` as tree-scoped references
https://bugs.webkit.org/show_bug.cgi?id=296048
rdar://156484228

Reviewed by Antti Koivisto.

Adopt `Style::Scope::resolveTreeScopedReference()` introduced in 300333@main to resolve `animation-name`
and find the correct scope to find the matching `@keyframes` rule.

Because `resolveTreeScopedReference()` provides a const `Style::Scope` and `Style::Resolver` to work with,
a few methods on `Style::Resolver` needed to be updated to be marked const, which they should have been
all along.

Additionally, because the `resolveTreeScopedReference()` lambda needs to return a truthy value to determine
whether a matching `@keyframes` rule was found for the provided scope, we needed to change
`Style::Resolver::keyframeStylesForAnimation()` to return a `bool` to indicate whether it processed an
`@keyframes` rule.

Tests: imported/w3c/web-platform-tests/css/css-animations/animation-name-in-nested-shadow.html
       imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-inner-match.html
       imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-outer-match.html
       imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part.html

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-nested-shadow-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-nested-shadow-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-nested-shadow.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-inner-match-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-inner-match.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-outer-match-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-outer-match.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-name-in-shadow-part.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/keyframes-002-expected.txt:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::computeCSSAnimationBlendingKeyframes):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::builderContext const):
(WebCore::Style::Resolver::styleForKeyframe const):
(WebCore::Style::Resolver::isAnimationNameValid const):
(WebCore::Style::Resolver::keyframeStylesForAnimation const):
(WebCore::Style::Resolver::builderContext): Deleted.
(WebCore::Style::Resolver::styleForKeyframe): Deleted.
(WebCore::Style::Resolver::isAnimationNameValid): Deleted.
(WebCore::Style::Resolver::keyframeStylesForAnimation): Deleted.
* Source/WebCore/style/StyleResolver.h:
* Source/WebCore/style/Styleable.cpp:
(WebCore::keyframesRuleExistsForAnimation):

Canonical link: https://commits.webkit.org/300706@main
